pallets/ddc-payouts/src/lib.rs

```diff
@@ -0,0 +1,720 @@
+			amount: u128,
+		},
+		ChargeFailed {
+			cluster_id: ClusterId,
+			era: DdcEra,
+			customer_id: T::AccountId,
+			amount: u128,
+		},
+		ChargingFinished {
+			cluster_id: ClusterId,
+			era: DdcEra,
+		},
+		RewardingStarted {
+			cluster_id: ClusterId,
+			era: DdcEra,
+		},
+		Rewarded {
+			cluster_id: ClusterId,
+			era: DdcEra,
+			node_provider_id: T::AccountId,
+			amount: u128,
+		},
+		RewardingFinished {
+			cluster_id: ClusterId,
+			era: DdcEra,
+		},
+		BillingReportFinalized {
+			cluster_id: ClusterId,
+			era: DdcEra,
+		},
+	}
+
+	#[pallet::error]
+	#[derive(PartialEq)]
+	pub enum Error<T> {
+		BillingReportDoesNotExist,
+		NotExpectedState,
+		Unauthorised,
+		BatchIndexAlreadyProcessed,
+		BatchIndexIsOutOfRange,
+		BatchesMissed,
+		NotDistributedBalance,
+		BatchIndexOverflow,
+		BoundedVecOverflow,
+		ArithmeticOverflow,
+		NotExpectedClusterState,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn active_billing_reports)]
+	pub type ActiveBillingReports<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		ClusterId,
+		Blake2_128Concat,
+		DdcEra,
+		BillingReport<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn authorised_caller)]
+	pub type AuthorisedCaller<T: Config> = StorageValue<_, T::AccountId>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn debtor_customers)]
+	pub type DebtorCustomers<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		ClusterId,
+		Blake2_128Concat,
+		T::AccountId,
+		u128,
+		ValueQuery,
+	>;
+
+	#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo, PartialEq)]
+	#[scale_info(skip_type_params(T))]
+	pub struct BillingReport<T: Config> {
+		state: State,
+		vault: T::AccountId,
+		total_customer_charge: CustomerCharge,
+		total_distributed_reward: u128,
+		total_node_usage: NodeUsage,
+		// stage 1
+		charging_max_batch_index: BatchIndex,
+		charging_processed_batches: BoundedBTreeSet<BatchIndex, MaxBatchesCount>,
+		// stage 2
+		rewarding_max_batch_index: BatchIndex,
+		rewarding_processed_batches: BoundedBTreeSet<BatchIndex, MaxBatchesCount>,
+	}
+
+	impl<T: pallet::Config> Default for BillingReport<T> {
+		fn default() -> Self {
+			Self {
+				state: State::default(),
+				vault: T::PalletId::get().into_account_truncating(),
+				total_customer_charge: CustomerCharge::default(),
+				total_distributed_reward: Zero::zero(),
+				total_node_usage: NodeUsage::default(),
+				charging_max_batch_index: Zero::zero(),
+				charging_processed_batches: BoundedBTreeSet::default(),
+				rewarding_max_batch_index: Zero::zero(),
+				rewarding_processed_batches: BoundedBTreeSet::default(),
+			}
+		}
+	}
+
+	#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo, PartialEq, Default)]
+	pub enum State {
+		#[default]
+		NotInitialized,
+		Initialized,
+		ChargingCustomers,
+		CustomersCharged,
+		DeductingFees,
+		RewardingProviders,
+		ProvidersRewarded,
+		Finalized,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
```

Comment:
Since this is a new pallet, it's better not to leave so much tech debt behind. Please introduce a weights.rs module and place weights there then use them here.

Reply:
will be introduced in the next few days